cephadm: Make easy manager modules development

jmolmo opened the pull request:
Using a volume, the content  in _</usr/share/mgr>_ folder in manager containers is replaced by
the content existing in one host folder.

The source host can be set using the new parameter `shared_mgr_folder`

This make easy to develop/debug because it is possible to make modifications "on-flight" in the manager modules:

A typical use can be:
`cephadm bootstrap --mon-ip $mon_ip --allow-fqdn-hostname --skip-dashboard --shared_mgr_folder /mnt/MyCode/ceph/src/pybind/mgr`

After the cluster is created, you can make modifications in your source folder: /mnt/MyCode/ceph/src/pybind/mgr
for example in the cephadm manager module, and after that test the modification just disabling and re-enabling the module again.

Note 1:
We have a couple of manager modules that need a special procedure in order to be able to modify code "on-flight":

- Dashboard: 
You will need to have generated the folder dist with the frontend components in your source folder

- Rook: 
Needed to install manually in the manager a rpm with  "rook shared" libraries. use for example:
```
https://download.ceph.com/rpm-octopus/el8/noarch/
ceph-mgr-rook-15.2.2-0.el8.noarch.rpm 
```

Probably we have room for more improvements ... any feedback will be welcome.

Note 2: 
Thanks @alfonsomthd for the idea and for the help testing! One (or more) Vermouth pending ...

Signed-off-by: Juan Miguel Olmo Martínez <jolmomar@redhat.com>

sebastian-philipp commented:
cc @jdurgin . Looks a bit like your cstart changes, right?

jmolmo commented:
> cc @jdurgin . Looks a bit like your cstart changes, right?

I think the main advantage of this change is that it can be used to debug/develop in a cluster with several vms used for different purposes. This kind of environment is more similar to the one that probably is going to be used in real envs. and I really like this part :-).

alfonsomthd commented:
@jmolmo Should this new option be documented in developer documentation?

sebastian-philipp commented:
@jan--f , I don't think this PR helps you developing c-v as it is, but maybe we can enhance this PR a bit?

sebastian-philipp commented on src/cephadm/cephadm:
```diff
@@ -4513,6 +4526,11 @@ def _get_parser():
         action='store_true',
         help='Do not automatically provision monitoring stack (prometheus, grafana, alertmanager, node-exporter)')
 
+    parser_bootstrap.add_argument(
+        '--shared_mgr_folder',
+        metavar='MGR_SOURCE_FOLDER',
+        help='Development mode. "/usr/share/ceph/mgr" in containers is a volume mapped to the manager modules source folder')
```
what about changing the base location to be the ceph source root directory? that would enable us the use the same flag name for different containers as well? (i.e. ceph-volume)

jmolmo replied:
Is a good idea!. I will check to see if we have any kind of unexpected behaviour, and if it is ok I will make the update.

jdurgin commented:
> > cc @jdurgin . Looks a bit like your cstart changes, right?

Yes, this is similar. I was also getting /usr/share/mgr from the source tree: https://gist.github.com/jdurgin/3215df1295d7bfee0e91b203eae71dce#file-prepare-dev-container-sh-L65

> I think the main advantage of this change is that it can be used to debug/develop in a cluster with several vms used for different purposes. This kind of environment is more similar to the one that probably is going to be used in real envs. and I really like this part :-).

Agreed, getting the dev environment closer to real deployments is great!

sebastian-philipp commented:
http://pulpito.ceph.com/swagner-2020-05-29_10:28:21-rados:cephadm-wip-swagner-testing-2020-05-28-1651-distro-basic-smithi/

sebastian-philipp commented:
```
2020-05-29T15:19:03.677 INFO:teuthology.orchestra.run.smithi149.stderr:INFO:cephadm:Creating mgr...
2020-05-29T15:19:03.683 INFO:teuthology.orchestra.run.smithi149.stderr:Traceback (most recent call last):
2020-05-29T15:19:03.683 INFO:teuthology.orchestra.run.smithi149.stderr:  File "/home/ubuntu/cephtest/cephadm", line 4739, in <module>
2020-05-29T15:19:03.684 INFO:teuthology.orchestra.run.smithi149.stderr:    r = args.func()
2020-05-29T15:19:03.684 INFO:teuthology.orchestra.run.smithi149.stderr:  File "/home/ubuntu/cephtest/cephadm", line 1157, in _default_image
2020-05-29T15:19:03.684 INFO:teuthology.orchestra.run.smithi149.stderr:    return func()
2020-05-29T15:19:03.684 INFO:teuthology.orchestra.run.smithi149.stderr:  File "/home/ubuntu/cephtest/cephadm", line 2532, in command_bootstrap
2020-05-29T15:19:03.684 INFO:teuthology.orchestra.run.smithi149.stderr:    mgr_c = get_container(fsid, 'mgr', mgr_id)
2020-05-29T15:19:03.684 INFO:teuthology.orchestra.run.smithi149.stderr:  File "/home/ubuntu/cephtest/cephadm", line 1731, in get_container
2020-05-29T15:19:03.685 INFO:teuthology.orchestra.run.smithi149.stderr:    volume_mounts=get_container_mounts(fsid, daemon_type, daemon_id),
2020-05-29T15:19:03.685 INFO:teuthology.orchestra.run.smithi149.stderr:  File "/home/ubuntu/cephtest/cephadm", line 1629, in get_container_mounts
2020-05-29T15:19:03.685 INFO:teuthology.orchestra.run.smithi149.stderr:    if os.path.exists(mgr_folder):
2020-05-29T15:19:03.685 INFO:teuthology.orchestra.run.smithi149.stderr:UnboundLocalError: local variable 'mgr_folder' referenced before assignment
```

jmolmo commented:
**Modifications:**

- Changed the parameter name.

In order to use this feature the command needed is:
`cephadm bootstrap --mon-ip <ip>  --shared_ceph_folder <ceph_folder>`

Where <ceph_folder> = **your ceph source code folder.**

- I have included other mounted folders to make easy also the development with "ceph-volume", and with dashboard.

- Fixed the error when this new parameter is not used.

**Two Warns: ( ideas to avoid this are welcomed!!)**

- Dashboard development:
Is needed to generate the java frontend assets in your dashboard source code folder.

- Rook orchestrator:
Need to install manually the ceph rook-client library. You need to install the rpm from ceph-mgr-rook-15.2.2-0.el8.noarch.rpm  or generate and  copy manually the library (https://github.com/ceph/rook-client-python) to the manager container.


**Target:**
My idea is to combine this with a vm generator to deploy a complete development environment ("n" vms) in less than "n" minutes.

sebastian-philipp commented:
http://pulpito.ceph.com/swagner-2020-06-02_12:46:10-rados:cephadm-wip-swagner3-testing-2020-06-02-1145-distro-basic-smithi/
